jcl/src/java.base/share/classes/openj9/internal/foreign/abi/LayoutStrPreprocessor.java

```diff
@@ -209,38 +215,151 @@ private static StringBuilder preprocessLayout(MemoryLayout targetLayout, boolean
+		int structElementCount = elementLayoutList.size();
+		StringBuilder elementLayoutStrs = new StringBuilder("");
+		int paddingElements = 0;
+
+		for (int elemIndex = 0; elemIndex < structElementCount; elemIndex++) {
+			MemoryLayout structElement = elementLayoutList.get(elemIndex);
+			/* Ignore any padding element in the struct as it is handled by ffi_call by default. */
+			if (isPaddingLayout(structElement)) {
+				paddingElements += 1;
+				/* The padding bytes is required in the native signature for upcall thunk generation. */
+				if (!isDownCall) {
+					elementLayoutStrs.append('(').append(structElement.byteSize()).append(')');
 				}
+			} else {
+				elementLayoutStrs.append(preprocessLayout(structElement, isDownCall));
 			}
-			/* Prefix "#" to denote the start of this layout string in the case of downcall. */
-			if (isDownCall) {
-				targetLayoutString.append('#').append(structElementCount - paddingElements);
+		}
+
+		/* Prefix "#" to denote the start of this layout string in the case of downcall. */
+		if (isDownCall) {
+			targetLayoutStr.append('#').append(structElementCount - paddingElements);
+		}
+		targetLayoutStr.append('[').append(elementLayoutStrs).append(']');
+
+		return targetLayoutStr;
+	}
+
+	/*[IF JAVA_SPEC_VERSION >= 21]*/
+	/* Encode the types in the union layout to a symbol string to simplify the further processing in native. */
+	private static StringBuilder encodeUnionLayoutStr(UnionLayout unionLayout, StringBuilder targetLayoutStr, boolean isDownCall) {
+		long unionByteSize = unionLayout.byteSize();
+		StringBuilder elementLayoutStrs = new StringBuilder("");
+		int byteAlignment = (int)unionLayout.byteAlignment();
+		ValueLayout sequenceElement = null;
+
+		/* The idea is to select the primitve type with the biggest alignement of the union
+		 * (obtained via unionLayout.byteAlignment()) to be laid out as a sequence which is
+		 * correctly mappped onto the given GPRs/FPRs on the hardware.
+		 */
+		switch (byteAlignment) {
+		case 1: /* 1 Byte. */
+			sequenceElement = JAVA_BYTE;
+			break;
+		case 2: /* 2 Bytes. */
+			sequenceElement = JAVA_SHORT;
+			break;
+		case 4: /* 4 Bytes. */
+			if (validatePrimTypesOfUnionForGPR(unionLayout, byteAlignment)) {
+				sequenceElement = JAVA_INT;
+			} else {
+				sequenceElement = JAVA_FLOAT;
 			}
-			targetLayoutString.append('[').append(elementLayoutStrs).append(']');
+			break;
+		case 8: /* 8 Bytes. */
+			if (validatePrimTypesOfUnionForGPR(unionLayout, byteAlignment)) {
+				sequenceElement = JAVA_LONG;
+			} else {
+				sequenceElement = JAVA_DOUBLE;
+			}
+			break;
+		default:
+			throw new IllegalArgumentException("Invalid alignment of the union layout: " + unionLayout);
 		}
 
-		return targetLayoutString;
+		long elementByteSize = sequenceElement.byteSize();
+		long sequenceCount = (0 == (unionByteSize % elementByteSize)) ?
+				(unionByteSize / elementByteSize) : (unionByteSize / elementByteSize + 1);
```

Comment:
Suggest extracting the common subexpression:
```java
		long sequenceCount = (unionByteSize / elementByteSize)
				+ ((0 == (unionByteSize % elementByteSize)) ? 0 : 1);
```

Reply:
Fixed as suggested above.